modules/auction_manager/src/lib.rs

```diff
@@ -112,10 +133,134 @@ decl_module! {
 		const GetStableCurrencyId: CurrencyIdOf<T> = T::GetStableCurrencyId::get();
 		const GetNativeCurrencyId: CurrencyIdOf<T> = T::GetNativeCurrencyId::get();
 		const GetAmountAdjustment: Rate = T::GetAmountAdjustment::get();
+
+		pub fn cancel(origin, id: AuctionIdOf<T>) {
+			ensure_none(origin)?;
+			ensure!(Self::is_shutdown(), Error::<T>::MustAfterShutdown);
+			<Module<T> as AuctionManager<T::AccountId>>::cancel_auction(id)?;
+		}
+
+		// Runs after every block.
+		fn offchain_worker(now: T::BlockNumber) {
+			if Self::is_shutdown() {
+				if let Err(e) = Self::_offchain_worker(now) {
+					debug::info!(
+						target: "auction-manager offchain worker",
+						"cannot run offchain worker at {:?}: {:?}",
+						now,
+						e,
+					);
+				}
+			}
+		}
 	}
 }
 
 impl<T: Trait> Module<T> {
+	fn acquire_offchain_worker_lock() -> Result<Timestamp, OffchainErr> {
```

Comment:
we should be able to abstract the offchain worker lock logic into orml utilities to avoid code duplications